squidpy/gr/_ppatterns.py

```diff
@@ -471,3 +507,109 @@ def _find_min_max(spatial: np.ndarray) -> Tuple[float, float]:
     ].astype(fp)
 
     return thres_min, thres_max
+
+
+def _p_value_calc(
+    score: np.ndarray,
+    sims: Union[np.ndarray, None],
+    weights: Union[spmatrix, np.ndarray],
+    params: Dict[str, Any],
+) -> Tuple[np.ndarray, float, np.ndarray, np.ndarray, Union[Any, np.ndarray]]:
+    """
+    Handle p-value calculation for spatial autocorrelation function.
+
+    Parameters
+    ----------
+    score
+        (n_features,).
+    sims
+        (n_simulations, n_features).
+    params
+        Object to store relevant function parameters.
+
+    Returns
+    -------
+    pval_norm
+        p-value under normality assumption
+    pval_sim
+        p-values based on permutations
+    pval_z_sim
+        p-values based on standard normal approximation from permutations
+
+    """
+    if sims is not None:
+        n_perms = sims.shape[0]
+        large_perm = (sims >= score).sum(axis=0)
+        # subtract total perm for negative values
+        large_perm[(n_perms - large_perm) < large_perm] = n_perms - large_perm[(n_perms - large_perm) < large_perm]
+        # get p-value based on permutation
+        p_sim: np.ndarray = (large_perm + 1) / (n_perms + 1)
+
+        # get p-value based on standard normal approximation from permutations
+        e_score_sim = sims.sum(axis=0) / n_perms
+        se_score_sim = sims.std(axis=0)
+        z_sim = (score - e_score_sim) / se_score_sim
+        p_z_sim = np.empty(z_sim.shape)
+
+        p_z_sim[z_sim > 0] = 1 - stats.norm.cdf(z_sim[z_sim > 0])
+        p_z_sim[z_sim <= 0] = stats.norm.cdf(z_sim[z_sim <= 0])
+
+        var_sim = np.var(sims, axis=0)
+    else:
+        p_sim = p_z_sim = var_sim = np.empty(score.shape[0]) * np.nan
+
+    p_norm, var_norm = _analytic_pval(score, weights, params)
```

Comment:
If a division by 0 can occur in this function, maybe consider wrapping the call in https://numpy.org/doc/stable/reference/generated/numpy.errstate.html
Not necesary though.